ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +377,296 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
+   * (added this logic to have mandatory sleep as we will not try to create new groups after this and will not check
+   * if all replicas are caught up)
+   * @param inflightRemoteReplicaGroup {@link Map}
+   * @param groupIdIterationCountMap {@link Map}
+   * @param allReplicasCaughtUpEarly {@link MutableBoolean}
+   * @param maxIterationsReached {@link MutableBoolean} will be set to true if max iteration limit reached
+   * @return {@link boolean} returns true if new groups can be created, false otherwise
+   */
+  private boolean shouldCreateNewRemoteReplicaGroups(Map<Integer, RemoteReplicaGroup> inflightRemoteReplicaGroup,
+      Map<Integer, Integer> groupIdIterationCountMap, MutableBoolean allReplicasCaughtUpEarly,
+      MutableBoolean maxIterationsReached) {
+
+    //check if thread is in shutdown or cycle termination is received
+    if (!running || shouldTerminateCurrentCycle) {
+      logger.trace(
+          "Thread name: {} not creating new groups as thread is in shutdown mode {} or terminate cycle received {}",
+          threadName, !running, shouldTerminateCurrentCycle);
+      return false;
+    }
+
+    //get max iterations by any remote replica group
+    int maxIterationCount = 0;
+    for (Map.Entry<Integer, Integer> entry : groupIdIterationCountMap.entrySet()) {
+      maxIterationCount = Integer.max(maxIterationCount, entry.getValue());
+    }
+
+    //check if any replica reached max iteration count
+    if (maxIterationCount == maxIterationsPerGroupPerCycle) {
+      if (maxIterationsReached.isFalse()) {
+        maxIterationsReached.setTrue();
+        if (inflightRemoteReplicaGroup.isEmpty()) {
+          allReplicasCaughtUpEarly.setTrue();
+          logger.trace("Thread name: {} All replicas marked to caught up early", threadName);
+        }
+      }
+      logger.trace("Thread name: {} not creating new groups as iteration limit {} reached", threadName,
+          maxIterationsPerGroupPerCycle);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Generates active and stand by replicas with no progress,
+   * checks and filters replicas which are throttled and are present in any iteration group
+   * @param activeReplicas {@link List}
+   * @param standbyReplicasWithNoProgress {@link List}
+   * @param remoteReplicas {@link List}
+   * @param remoteReplicaToThrottledTill {@link Map}
+   * @param standbyReplicasNoProgressReplicaGroupId {@link Integer}
+   * @param groupId {@link Integer}
+   * @param inflightRemoteReplicaGroup {@link Map}
+   */
+  private void getActiveAndStandByNoProgressReplicas(List<RemoteReplicaInfo> activeReplicas,
+      List<RemoteReplicaInfo> standbyReplicasWithNoProgress, List<RemoteReplicaInfo> remoteReplicas,
+      Map<RemoteReplicaInfo, Long> remoteReplicaToThrottledTill, Integer standbyReplicasNoProgressReplicaGroupId,
+      Integer groupId, Map<Integer, RemoteReplicaGroup> inflightRemoteReplicaGroup) {
```

Comment:
let's use primitive type for these two integers.

Reply:
changed and refactored.